### DISCUSSION TITLE
Access GPIO TX RX serial ports ?

### DISCUSSION THREAD

**Discussion by raydacosta** [Pi4J/pi4j#348] [General]
Access GPIO TX RX serial ports

I'm trying to read the RFID data being sent via the 125khz Rfid Reader Module + Antenam using the PI4J V2 but without success. What could I be doing wrong?

`public void listenerTXRX() {
        try {

            //I don't know exactly what this variable does
            String SERIAL_ADDRESS = "/dev/ttyS0";
            var console = new Console();
            
            var serial = pi4j.create(Serial.newConfigBuilder(pi4j)
                    .use_9600_N81()
                    .dataBits_8()
                    .parity(Parity.NONE)
                    .stopBits(StopBits._1)
                    .flowControl(FlowControl.NONE)
                    .id("my-serial")
                    .device(SERIAL_ADDRESS)
                    //.provider("pigpio-serial")
                    .build());
            serial.open();

            // Wait till the serial port is open
            console.print("Esperando até que a porta serial esteja aberta");
            while (!serial.isOpen()) {
                Thread.sleep(250);
            }

             // Start a thread to handle the incoming data from the serial port
            SerialReader serialReader = new SerialReader(serial);
            Thread serialReaderThread = new Thread(serialReader, "SerialReader");
            serialReaderThread.setDaemon(true);
            serialReaderThread.start();

            while (serial.isOpen()) {
                Thread.sleep(500);
            }

            serialReader.stopReading();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }`

Keep stopping at this place

`console.print("Waiting till serial port is open");
while (!serial.isOpen()) {
    Thread.sleep(250);
}
![WhatsApp Image 2024-04-16 at 23 51 14](https://github.com/Pi4J/pi4j-v2/assets/32819426/17fb8b0b-95a4-4615-96fe-a7564fbe98c4)
`

**Comment by taartspi** [comment]
You commented out the provider statement. This is needed.   That address is a path.  U should be able to ls /dev and see the entries

**Reply by raydacosta** [reply]
Hi taartspi, thank you very much for your support. Just to make sure. So you're telling me that no: .provider("pigpio-serial") must have the address of the port I'm opening.: Example: .provider("/dev/ttyS0") or .provider("ttyS0"). Right?

**Reply by raydacosta** [reply]
Another help, I am connected to GPIO pins 14 and 15. How can I find out which port it is?

**Reply by raydacosta** [reply]
I made the changes, but I still have the problem, I must be doing something wrong. Is there any example on the internet that you could recommend with PI4J V2?

**/dev/ttyS0**
`PLATFORMS: [1] "Pi4J Runtime Platforms" <com.pi4j.platform.impl.DefaultPlatforms>
└─PLATFORM: "RaspberryPi Platform" {raspberrypi} <com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform> {Pi4J Platform for the RaspberryPi series of products.}
com.pi4j.provider.exception.ProviderNotFoundException: Pi4J provider [/dev/ttyS0] could not be found.  Please include this 'provider' JAR in the classpath.
        at com.pi4j.provider.impl.DefaultRuntimeProviders.get(DefaultRuntimeProviders.java:227)
        at com.pi4j.provider.impl.DefaultProviders.get(DefaultProviders.java:147)
        at com.pi4j.provider.Providers.get(Providers.java:253)
        at com.pi4j.context.Context.create(Context.java:327)
        at com.pi4j.internal.IOCreator.create(IOCreator.java:58)
        at com.pi4j.internal.IOCreator.create(IOCreator.java:146)
        at com.wrapper.wrapperpointpi.WrapperPointPI.listenerTXRX(WrapperPointPI.java:301)
        at com.wrapper.wrapperpointpi.WrapperPointPI.main(WrapperPointPI.java:343)
[main] INFO com.pi4j.runtime.impl.DefaultRuntime - Shutting down Pi4J context/runtime...
[main] INFO com.pi4j.util.ExecutorPool - Shutting down executor pool Pi4J.RUNTIME
[main] INFO com.pi4j.runtime.impl.DefaultRuntime - Pi4J context/runtime successfully shutdown. Dispatching shutdown event.
wrapper@raspberrypi:~ $




**ttyS0**
`PLATFORMS: [1] "Pi4J Runtime Platforms" <com.pi4j.platform.impl.DefaultPlatforms>
└─PLATFORM: "RaspberryPi Platform" {raspberrypi} <com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform> {Pi4J Platform for the RaspberryPi series of products.}
com.pi4j.provider.exception.ProviderNotFoundException: Pi4J provider [ttyS0] could not be found.  Please include this 'provider' JAR in the classpath.
        at com.pi4j.provider.impl.DefaultRuntimeProviders.get(DefaultRuntimeProviders.java:227)
        at com.pi4j.provider.impl.DefaultProviders.get(DefaultProviders.java:147)
        at com.pi4j.provider.Providers.get(Providers.java:253)
        at com.pi4j.context.Context.create(Context.java:327)
        at com.pi4j.internal.IOCreator.create(IOCreator.java:58)
        at com.pi4j.internal.IOCreator.create(IOCreator.java:146)
        at com.wrapper.wrapperpointpi.WrapperPointPI.listenerTXRX(WrapperPointPI.java:301)
        at com.wrapper.wrapperpointpi.WrapperPointPI.main(WrapperPointPI.java:343)
[main] INFO com.pi4j.runtime.impl.DefaultRuntime - Shutting down Pi4J context/runtime...
[main] INFO com.pi4j.util.ExecutorPool - Shutting down executor pool Pi4J.RUNTIME
[main] INFO com.pi4j.runtime.impl.DefaultRuntime - Pi4J context/runtime successfully shutdown. Dispatching shutdown event.




`


`

**Comment by taartspi** [comment]
var serial = pi4j.create(Serial.newConfigBuilder(pi4j)
                .use_9600_N81()
                .dataBits_8()
                .parity(Parity.NONE)
                .stopBits(StopBits._1)
                .flowControl(FlowControl.NONE)
                .id("my-serial")
                .device(SERIAL_ADDRESS)
                //.provider("pigpio-serial")
                .build());
        serial.open();  
Uncomment the line   //.provider("pigpio-serial")         Don't change it, just uncomment it.      
  
I wasn't certain on finding the address so I used google to find instructions.
On my machine
  ls -l  /dev/serial0
lrwxrwxrwx 1 root root 8 Apr 18 11:34 /dev/serial0 -> ttyAMA10    
Then looked in /dev and found it.
So the address is    /dev/ttyAMA10

**Reply by raydacosta** [reply]
Ok taartspi, I'll change and test your approach. Thanks for the feedback.

**Reply by raydacosta** [reply]
wrapper@raspberrypi:~ $ ls -l /dev/serial0
lrwxrwxrwx 1 root root 5 abr 18 14:13 /dev/serial0 -> ttyS0

**Comment by taartspi** [comment]
and be certain you enabled the serial port interface in raspi-config

**Reply by raydacosta** [reply]
Ok.. 

I had already done these procedures and made sure it was enabled;

$sudo raspi-config
Navigate to "Interfacing Options" > "Serial" and then choose "No" to disable the terminal on the serial port.

$sudo nano /boot/config.txt

Add the following line at the end of the file: enable_uart=1
$sudo nano /boot/cmdline.txt
Remove any mention of console=serial0,115200 or console=ttyAMA0,115200. Save and close the file.

$sudo reboot

**Reply by raydacosta** [reply]
Oops, it has evolved, I believe it is now a basic thing. Because it is listening to port pin 15.
After making sure that all of the above settings have been applied. Hi, apparently you are now reading the GPIO. I am receiving data from the GPIO, because if I remove the pins, it stops sending data.
I admit that I am now using the wrong implementation.


`public void startListenrSerial() {
        try {

            String SERIAL_ADDRESS = "/dev/ttyS0";

            open();

            System.out.println("-------------------------------------------------");
            System.out.println("PI4J PROVIDERS");
            System.out.println("-------------------------------------------------");
            pi4j.providers().describe().print(System.out);
            System.out.println("-------------------------------------------------");

            Console console = new Console();
            console.print("Porta: " + SERIAL_ADDRESS);
            Serial serial = pi4j.create(Serial.newConfigBuilder(pi4j)
                    .use_115200_N81()
                    .dataBits_7()
                    .parity(Parity.NONE)
                    .stopBits(StopBits._1)
                    .flowControl(FlowControl.NONE)
                    .id("my-serial")
                    .device(SERIAL_ADDRESS)
                    .provider("pigpio-serial")
                    .build());
            serial.open();

            // Wait till the serial port is open
            console.print("Aguardando abrir a porta");
            while (!serial.isOpen()) {
                Thread.sleep(250);
            }

            console.print("Iniciando listener.....");
            // Start a thread to handle the incoming data from the serial port
            SerialReader serialReader = new SerialReader(serial);
            Thread serialReaderThread = new Thread(serialReader, "SerialReader");
            serialReaderThread.setDaemon(true);
            serialReaderThread.start();

            while (serial.isOpen()) {
                Thread.sleep(500);
            }

            serialReader.stopReading();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
`
The problem now is that you are reading all the characters wrong.

` @Override
    public void run() {

        // We use a buffered reader to handle the data received from the serial port
        BufferedReader br = new BufferedReader(new InputStreamReader(serial.getInputStream()));

        try {
            // Data from the GPS is recieved in lines
            String line = "";

            // Read data until the flag is false
            while (continueReading) {
                // First we need to check if there is data available to read.
                // The read() command for pigio-serial is a NON-BLOCKING call, 
                // in contrast to typical java input streams.
                var available = serial.available();
                if (available > 0) {
                    for (int i = 0; i < available; i++) {
                        byte b = (byte) br.read();
                        if (b < 32) {
                            // All non-string bytes are handled as line breaks
                            if (!line.isEmpty()) {
                                // Here we should add code to parse the data to a GPS data object
                                System.out.println("Data: '" + line + "'");
                                line = "";
                            }
                        } else {
                            line += (char) b;
                        }
                    }
                } else {
                    Thread.sleep(10);
                }
            }
        } catch (Exception e) {
            System.out.println(e.getStackTrace());
        }
    }`


Resust:
`VERSION: WrapperUSB-1.0.0.3
=============================================
-------------------------------------------------
PI4J PROVIDERS
-------------------------------------------------
PROVIDERS: [6] "I/O Providers" <com.pi4j.provider.impl.DefaultProviders>
├─DIGITAL_OUTPUT: [1] <com.pi4j.io.gpio.digital.DigitalOutputProvider>
│ └─PROVIDER: "PiGpio Digital Output (GPIO) Provider" {pigpio-digital-output} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProviderImpl}
├─I2C: [1] <com.pi4j.io.i2c.I2CProvider>
│ └─PROVIDER: "PiGpio I2C Provider" {pigpio-i2c} <com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProviderImpl> {com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProviderImpl}
├─SPI: [1] <com.pi4j.io.spi.SpiProvider>
│ └─PROVIDER: "PiGpio SPI Provider" {pigpio-spi} <com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl> {com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProviderImpl}
├─ANALOG_INPUT: [0] <com.pi4j.io.gpio.analog.AnalogInputProvider>
├─SERIAL: [1] <com.pi4j.io.serial.SerialProvider>
│ └─PROVIDER: "PiGpio Serial Provider" {pigpio-serial} <com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl> {com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProviderImpl}
├─ANALOG_OUTPUT: [0] <com.pi4j.io.gpio.analog.AnalogOutputProvider>
├─DIGITAL_INPUT: [1] <com.pi4j.io.gpio.digital.DigitalInputProvider>
│ └─PROVIDER: "PiGpio Digital Input (GPIO) Provider" {pigpio-digital-input} <com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl> {com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProviderImpl}
└─PWM: [1] <com.pi4j.io.pwm.PwmProvider>
  └─PROVIDER: "PiGpio PWM Provider" {pigpio-pwm} <com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProviderImpl> {com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProviderImpl}
-------------------------------------------------
[main] INFO com.pi4j.util.Console - Porta: /dev/ttyS0
[main] INFO com.pi4j.util.Console - Aguardando abrir a porta
[main] INFO com.pi4j.util.Console - Iniciando listener.....
[SerialReader] WARN com.pi4j.library.pigpio.impl.PiGpioNativeImpl - PIGPIO ERROR: PI_SER_READ_NO_DATA; ser read no data available

**Reply by raydacosta** [reply]
Result dados:

`Data: '@'
Data: '`'
Data: ' '
Data: :'
Data: 'D@'
Data: '?'
[SerialReader] WARN com.pi4j.library.pigpio.impl.PiGpioNativeImpl - PIGPIO ERROR: PI_SER_READ_NO_DATA; ser read no data available
[SerialReader] WARN com.pi4j.library.pigpio.impl.PiGpioNativeImpl - PIGPIO ERROR: PI_SER_READ_NO_DATA; ser read no data available
Data: 'raspberrypi login: '
Data: 'raspberrypi login: '
Data: 'raspberrypi login: '
Data: 'raspberrypi login: Debian GNU/Linux 12 raspberrypi ttyS0'
`

`

**Comment by taartspi** [comment]
I see data in the printlines with what appears to be the login prompt.  I think you need to research why as that ASCII is present as it will interfere with your expected data. I know you mentioned removing the serial console data from the cmdline.txt but it appears the OS still believes it is a console connection.       In the earlier printouts seeing Data six times you must have received 6 bytes of value less than 32 and a few displayable ASCII bytes,  But the fact you later see the login prompt all this data maybe invalid.  
You mentioned using a RFID reader and the run() thread refers to GPS.  What is the connected device and its part number.
